caproto/_data.py

```diff
@@ -219,60 +246,73 @@ def __init__(self, *, value=None, timestamp=None, status=0, severity=0,
-        return convert_values(values=values, from_dtype=native_from,
-                              to_dtype=self.data_type,
-                              string_encoding=self.string_encoding,
-                              enum_strings=getattr(self, 'enum_strings', None))
-
-    def astype(self, to_dtype):
-        '''Convenience function: convert stored data to a specific type'''
-        native_to = native_type(to_dtype)
-        return convert_values(values=self.value, from_dtype=self.data_type,
-                              to_dtype=native_to,
-                              string_encoding=self.string_encoding,
-                              enum_strings=getattr(self, 'enum_strings', None))
-
-    async def get_dbr_data(self, type_):
+    async def read(self, hostname, username, data_type):
         '''Get DBR data and native data, converted to a specific type'''
+        access = self.check_access(hostname, username)
+        if access not in (AccessRights.READ, AccessRights.READ_WRITE):
+            raise Forbidden("Client with hostname {} and username {} cannot "
+                            "read.".format(hostname, username))
         # special cases for alarm strings and class name
-        if type_ == ChType.STSACK_STRING:
-            ret = await self.alarm.get_dbr_data()
+        if data_type == ChType.STSACK_STRING:
+            ret = await self.alarm.read()
             return (ret, b'')
-        elif type_ == ChType.CLASS_NAME:
-            class_name = DBR_TYPES[type_]()
+        elif data_type == ChType.CLASS_NAME:
+            class_name = DBR_TYPES[data_type]()
             rtyp = self.reported_record_type.encode(self.string_encoding)
             class_name.value = rtyp
             return class_name, b''
 
         # for native types, there is no dbr metadata - just data
-        native_to = native_type(type_)
-        values = self.astype(native_to)
+        native_to = native_type(data_type)
+        values = convert_values(values=self.value, from_dtype=self.data_type,
+                              to_dtype=native_to,
+                              string_encoding=self.string_encoding,
+                              enum_strings=getattr(self, 'enum_strings', None))
+
 
-        if type_ in native_types:
+        if data_type in native_types:
             return b'', values
 
-        if type_ in self._dbr_metadata:
-            dbr_metadata = self._dbr_metadata[type_]
+        if data_type in self._dbr_metadata:
+            dbr_metadata = self._dbr_metadata[data_type]
         else:
             # TODO: non-standard type request. frequent ones probably should be
             # cached?
-            dbr_metadata = DBR_TYPES[type_]()
+            dbr_metadata = DBR_TYPES[data_type]()
 
         self._copy_metadata_to_dbr(dbr_metadata)
         return dbr_metadata, values
 
-    async def set_dbr_data(self, data, data_type, metadata):
+    async def write(self, hostname, username, data, data_type, metadata):
```

Comment:
Ditto - should be able to shove in data via python too without caring about CA access rights

Reply:
How about `auth_write` for authenticated writes by the client and `write` for unauthenticated direct access?